engine/access/ingestion/tx_error_messages/requester.go

```diff
@@ -0,0 +1,156 @@
+package tx_error_messages
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	execproto "github.com/onflow/flow/protobuf/go/flow/execution"
+	"github.com/rs/zerolog"
+	"github.com/sethvargo/go-retry"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/onflow/flow-go/engine/common/rpc/convert"
+	"github.com/onflow/flow-go/utils/logging"
+
+	"github.com/onflow/flow-go/engine/access/rpc/backend"
+	"github.com/onflow/flow-go/engine/common/rpc"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// RequesterConfig contains the retry settings for the tx error messages fetch.
+type RequesterConfig struct {
+	// the initial delay used in the exponential backoff for failed tx error messages download
+	// retries.
+	RetryDelay time.Duration
+	// the max delay used in the exponential backoff for failed tx error messages download.
+	MaxRetryDelay time.Duration
+}
+
+type Requester struct {
+	logger                     zerolog.Logger
+	config                     *RequesterConfig
+	backend                    *backend.Backend
+	execNodeIdentitiesProvider *rpc.ExecutionNodeIdentitiesProvider
+	executionResult            *flow.ExecutionResult
+}
+
+func NewRequester(
+	logger zerolog.Logger,
+	config *RequesterConfig,
+	backend *backend.Backend,
+	execNodeIdentitiesProvider *rpc.ExecutionNodeIdentitiesProvider,
+	executionResult *flow.ExecutionResult,
+) *Requester {
+	return &Requester{
+		logger:                     logger,
+		config:                     config,
+		backend:                    backend,
+		execNodeIdentitiesProvider: execNodeIdentitiesProvider,
+		executionResult:            executionResult,
+	}
+}
+
+// Request fetches transaction error messages for the specific
+// execution result this requester was configured with.
+//
+// No errors are expected during normal operations.
+func (r *Requester) Request(ctx context.Context) ([]flow.TransactionResultErrorMessage, error) {
+	backoff := retry.NewExponential(r.config.RetryDelay)
+	backoff = retry.WithCappedDuration(r.config.MaxRetryDelay, backoff)
+	backoff = retry.WithJitterPercent(15, backoff)
+
+	blockID := r.executionResult.BlockID
+	resultID := r.executionResult.ID()
+
+	var errorMessages []flow.TransactionResultErrorMessage
+
+	attempt := 0
+	err := retry.Do(ctx, backoff, func(context.Context) error {
+		if attempt > 0 {
+			r.logger.Debug().
+				Str("block_id", blockID.String()).
+				Str("result_id", resultID.String()).
+				Uint64("attempt", uint64(attempt)).
+				Msgf("retrying download")
+		}
+		attempt++
+
+		var err error
+		errorMessages, err = r.request(ctx, blockID, resultID)
+		if errors.Is(err, rpc.ErrNoENsFoundForExecutionResult) || status.Code(err) != codes.Canceled {
```

Comment:
I don't think this is quite right since `status.Code()` will wrap any error as `codes.Unknown`.
Instead you'll need to use `status.FromError()` an explicitly check that the error is a valid status code.

This is retryable if:
* `err` is `rpc.ErrNoENsFoundForExecutionResult`
* `err` is a valid grpc status code, and is not `codes.Canceled`

Any other error is not retryable